packages/peregrine/lib/store/actions/cart/asyncActions.js

```diff
@@ -134,51 +134,77 @@ export const addItemToCart = (payload = {}) => {
     };
 };
 
+/**
+ * @param payload.cartItemId {Number} the id of the cart item we are updating
+ * @param payload.item {Object} the new configuration item if changes are selected.
+ * TODO: Test for update without any changes
+ * TODO: Test for simple update without changes
+ */
 export const updateItemInCart = (payload = {}) => {
-    const { cartItemId, fetchCartId, item } = payload;
+    const {
+        cartItemId,
+        fetchCartId,
+        item,
+        productType,
+        quantity,
+        updateItem
+    } = payload;
     const writingImageToCache = writeImageToCache(item);
 
     return async function thunk(dispatch, getState) {
         await writingImageToCache;
         dispatch(actions.updateItem.request(payload));
 
-        const { cart, user } = getState();
+        const {
+            cart: { cartId },
+            user
+        } = getState();
         const { isSignedIn } = user;
 
         try {
-            let cartEndpoint;
-
-            if (!isSignedIn) {
-                const { cartId } = cart;
-
-                if (!cartId) {
-                    const missingCartIdError = new Error(
-                        'Missing required information: cartId'
-                    );
-                    missingCartIdError.noCartId = true;
-                    throw missingCartIdError;
-                }
-
-                cartEndpoint = `/rest/V1/guest-carts/${cartId}/items/${cartItemId}`;
+            if (productType === 'simple') {
+                // If the product is a simple product we can just use the
+                // updatecartItems graphql mutation.
+                await updateItem({
+                    variables: {
+                        cartId,
+                        itemId: cartItemId,
+                        quantity
+                    }
+                });
+                // and fetch details
+                await dispatch(
+                    getCartDetails({
+                        forceRefresh: true,
+                        fetchCartId
+                    })
+                );
             } else {
-                cartEndpoint = `/rest/V1/carts/mine/items/${cartItemId}`;
+                await dispatch(
+                    removeItemFromCart({
+                        item: {
+                            item_id: cartItemId
+                        },
+                        fetchCartId
+                    })
+                );
+                await dispatch(
+                    addItemToCart({
+                        ...payload,
+                        // For now, rewrite productType since the REST cart_item
+                        // uses "configurable" and "simple" string values.
+                        productType: 'ConfigurableProduct',
+                        fetchCartId
+                    })
+                );
```

Comment:
The above actions both call `getCartDetails`. The result is a weird double fetch. There's no way around this unless we break the calls to `getCartDetails` out of those functions. Probably worth moving it to the callback invoker as a "post action" action but not worth changing in this PR.

Reply:
I would be fine with this, but it does cause a weird flash of an empty cart, and then the add action finishes and displays the updated cart. I see that the `isUpdatingItem` state is being drilled down to these components, but `isLoading` seems to be causing some extra renders.

I tried changing some logic in `MiniCart/body`, but didn't work as I wanted; only displaying loading indicator `if (isLoading && !isUpdating)`. The only option I see is to defer these cart renders, but agree we should avoid a big re-factor in this scope. Any ideas to prevent that double render on update?